lib/aws/sns/uri.rb

```diff
@@ -0,0 +1,68 @@
+require "aws-sdk-sns"
+require "addressable/uri"
+
+module Aws
+  module SNS
+    class URI
+      attr_accessor :key_id, :access_key, :region, :topic_owner, :topic_name
+
+      def self.parse(uri)
+        return if uri.nil?
+
+        uri = Addressable::URI.parse(uri) unless uri.is_a?(Addressable::URI)
+
+        raise Aws::SNS::Errors::InvalidURIError unless uri.scheme == 'sns'
+
+        region = uri.host
+
+        if uri.user.blank? || uri.password.blank?
+          raise Aws::SNS::Errors::InvalidURIError, "User or Password missing"
+        end
+
+        key_id = CGI.unescape(uri.user)
+        access_key = CGI.unescape(uri.password)
+
+        segments = uri.path&.split('/')
+
+        raise Aws::SNS::Errors::InvalidURIError unless segments&.size == 3
+
+        topic_owner = CGI.unescape(segments[1])
+        topic_name = CGI.unescape(segments[2])
+
+        new(
+          key_id: key_id,
+          access_key: access_key,
+          region: region,
+          topic_owner: topic_owner,
+          topic_name: topic_name
+        )
+      end
+
+      def initialize(key_id:, access_key:, region:, topic_owner:, topic_name:)
+        @key_id = key_id
+        @access_key = access_key
+        @region = region
+        @topic_owner = topic_owner
+        @topic_name = topic_name
+      end
+
+      def arn
+        "arn:aws:sns:#{region}:#{topic_owner}:#{topic_name}"
+      end
+
+      def test_publish!
```

Comment:
Ruby newbie question: is the reason why you're adding "!" to the method name to hint that this method would mutate external state (e.g. send SNS to uri being tested?)

Reply:
The original meaning is mutate the object it's called on actually. Maybe I should not use it here actually. Good point!